src/transform/src/join_implementation.rs

```diff
@@ -126,21 +135,93 @@ impl JoinImplementation {
                     .into_iter()
                     .map(|typ| typ.keys)
                     .collect::<Vec<_>>();
                 let mut available_arrangements = vec![Vec::new(); inputs.len()];
+                let mut filters = Vec::new();
+
+                // We figure out what predicates from mfp_above could be pushed to which input.
+                // We won't actually push these down now; this just informs FilterCharacteristics.
+                let (map, mut filter, _) = mfp_above.as_map_filter_project();
+                let all_errors = filter.iter().all(|p| p.is_literal_err());
+                let (_, pushed_through_map) = PredicatePushdown::push_filters_through_map(
+                    &map,
+                    &mut filter,
+                    mfp_above.input_arity,
+                    all_errors,
+                )?;
+                let (_, push_downs) = PredicatePushdown::push_filters_through_join(
+                    &input_mapper,
+                    equivalences,
+                    pushed_through_map,
+                );
+
                 for index in 0..inputs.len() {
                     // We can work around mfps, as we can lift the mfps into the join execution.
                     let (mfp, input) =
                         MapFilterProject::extract_non_errors_from_expr(&inputs[index]);
-                    let (_, _, project) = mfp.as_map_filter_project();
-                    // Get and ArrangeBy expressions contribute arrangements.
+                    let (_, filter, project) = mfp.as_map_filter_project();
+
+                    // We gather filter characteristics:
+                    // - From the filter that is directly at the top mfp of the input.
+                    // - IndexedFilter joins are constructed from literal equality filters.
+                    // - If the input is an ArrangeBy, then we gather filter characteristics from
+                    //   the mfp below the ArrangeBy. (JoinImplementation often inserts ArrangeBys.)
+                    // - From filters that could be pushed down from above the join to this input.
+                    //   (In LIR, these will be executed right after the join path executes the join
+                    //   for this input.)
+                    let mut characteristics =
+                        FilterCharacteristics::filter_characteristics(&filter)?;
+                    if matches!(
+                        input,
+                        MirRelationExpr::Join {
+                            implementation: IndexedFilter(..),
+                            ..
+                        }
+                    ) {
+                        characteristics.add_literal_equality();
+                    }
+                    if let MirRelationExpr::ArrangeBy {
+                        input: arrange_by_input,
+                        ..
+                    } = input
+                    {
+                        let (mfp, input) =
```

Comment:
Yeah, I'm becoming more and more convinced nowadays that we should move all these arrangement things to LIR. Having these ArrangeBys in MIR is not so useful, because it's very hard to guess
- whether it's forming an arrangement right there or just a marker that that stuff is already arranged;
- whether the arrangement will be scanned or used in a lookup (https://github.com/MaterializeInc/materialize/issues/15242).

And then EXPLAINing indexes would become the responsibility of EXPLAIN PHYSICAL PLAN.

Reply:
AFAIU the main obstacle at the moment is the fact that certain MIR optimizations (such as `JoinOrdering`) impose a constraint that an input should be arranged, and if the input is consumed more than once in a `Join` node, a first run might come up with a differential plan that is guaranteed to ad-hoc create all arrangements that were missing to pick a delta plan.

> And then EXPLAINing indexes would become the responsibility of EXPLAIN PHYSICAL PLAN.

I think we can (and probably should) do this even now, and it will be much more robust if we do this kind of inference on top of LIR.